File: src/main/java/it/unibo/crabinv/Model/entities/enemies/wave/Wave.java
package it.unibo.crabinv.Model.entities.enemies.wave;

import it.unibo.crabinv.Model.entities.enemies.Enemy;

import java.util.List;

/**
 * Is the interface that has all the methods that the wave has
 */

public interface Wave {

    /**
     * Updates the Wave status by one tick.
     */
    void tickUpdate();

    /**
     * Gives the entire list of all the enemies that are alive:
     * @return a list of all enemies still alive
     */
    List<Enemy> getAliveEnemies();

    /**
     *Checks if the wave is finished
     * @return true if wave is finished, false if it isn't
     */
    boolean isWaveFinished();

    /**
     * Gets the maximum of enemies that are available to spawn
     * @return the number of available spawn slots for the enemies
     */
    int getMaxSpawnSlots();

}

File: src/main/java/it/unibo/crabinv/model/entities/enemies/wave/Wave.kt
package it.unibo.crabinv.model.entities.enemies.wave

import it.unibo.crabinv.model.entities.enemies.Enemy

/**
 * Is the interface that has all the methods that the wave has.
 */
interface Wave {
    /**
     * Updates the Wave status by one tick.
     */
    fun tickUpdate()

    /**
     * Gives the entire list of all the enemies that are alive.
     * @return a list of all enemies still alive
     */
    val aliveEnemies: MutableList<Enemy>

    /**
     * Checks if the wave is finished.
     * @return true if wave is finished, false if it isn't
     */
    val isWaveFinished: Boolean

    /**
     * Sets the Y coordinates spawn.
     * @param spawnY the new Y spawn coordinates
     */
    fun setSpawnY(spawnY: Double)
}
